api/src/Responder/Paginator/RangeRequest.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: redjik
+ * Date: 30.07.15
+ * Time: 14:58
+ */
+
+namespace Spira\Responder\Paginator;
+
+use Illuminate\Http\Request;
+
+class RangeRequest implements PaginatedRequestDecoratorInterface
+{
+    private $parsed = false;
+
+    private $offset = null;
+
+    private $limit = null;
+
+    private $isGetLast = false;
+
+    /**
+     * @var Request
+     */
+    private $request;
+
+    public function __construct(Request $request)
+    {
+        $this->request = $request;
+    }
+
+    /**
+     * @param null|int $default
+     * @return mixed
+     */
+    public function getOffset($default = null)
+    {
+        $this->parse();
+        return is_null($this->offset)?$default:$this->offset;
+    }
+
+    /**
+     * @param null|int $default
+     * @param null|int $max
+     * @return mixed
+     */
+    public function getLimit($default = null, $max = null)
+    {
+        $this->parse();
+        $limit = is_null($this->limit)?$default:$this->limit;
+        return ($limit > $max)?$max:$limit;
+    }
+
+    /**
+     * states that limit should be applied to the end of the result set
+     * @return bool
+     */
+    public function isGetLast()
+    {
+        $this->parse();
+        return $this->isGetLast;
+    }
+
+    /**
+     * @return bool
+     */
+    private function parse()
+    {
+        if ($this->parsed) {
+            return true;
+        }
+
+        $range = $this->getRequest()->headers?$this->getRequest()->headers->get('Range', ''):'';
```

Comment:
This ternary should be split out and have validation. This class represents a RangeRequest, and so if the request does not have ranges then it should be rejected with an exception (I think 400 Bad Request is most appropriate). It would make most sense for this exception to include the `Accept-Ranges` header so the client knows what type of Range header that they must request.

From the clients perspective it is unintuitive to make a GET all request and have only a subset of the data returned when they may have been expecting all. It is better for them to have a Bad Request exception notifying them that they must send appropriate range headers.